crypto/evp/keymgmt_lib.c

```diff
@@ -183,6 +200,124 @@ void *evp_keymgmt_fromdata(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
     return provdata;
 }
 
+int evp_keymgmt_is(const EVP_PKEY *pkey, int domainparams)
+{
+    EVP_KEYMGMT *keymgmt = pkey->pkeys[0].keymgmt;
+
+    if (keymgmt == NULL)
+        return 0;          /* There's no one to ask, so assume "no" */
+    return pkey->pkeys[0].domainparams
+        ? evp_keymgmt_isdomparams(keymgmt, pkey->pkeys[0].provdata)
+        : evp_keymgmt_iskey(keymgmt, pkey->pkeys[0].provdata);
+}
+
+/*
+ * This function adheres to EVP_PKEY_cmp() / EVP_PKEY_cmp_parameters()
+ * return values, i.e.:
+ *
+ *  1   same key
+ *  0   not same key
+ * -1   not same key type
+ * -2   unsupported operation
+ */
+int evp_keymgmt_cmp(const EVP_PKEY *pkey1, const EVP_PKEY *pkey2,
+                    int domainparams)
+{
+    EVP_KEYMGMT *keymgmt1 = pkey1->pkeys[0].keymgmt;
+    EVP_KEYMGMT *keymgmt2 = pkey2->pkeys[0].keymgmt;
+    void *provdata1 = NULL;
+    void *provdata2 = NULL;
+    void *exported_provdata2 = NULL;
+    int rv = 0;                  /* Assume the worst */
+
+    if (keymgmt1 == NULL && keymgmt2 == NULL)
+        return -2;   /* No support without at least one key manager */
+
+    if (keymgmt1 == NULL) {
+        EVP_KEYMGMT *swaptmp = keymgmt1;
+        keymgmt1 = keymgmt2;
+        keymgmt2 = swaptmp;
+    }
+
+    if (keymgmt1 != NULL && keymgmt2 != NULL
+        && EVP_KEYMGMT_number(keymgmt1) != EVP_KEYMGMT_number(keymgmt2))
+        return -1;               /* key type mismatch */
+
+    provdata1 = pkey1->pkeys[0].provdata;
+    provdata2 = pkey2->pkeys[0].provdata;
+
+    if (pkey1->pkeys[0].domainparams != domainparams
+        || pkey2->pkeys[0].domainparams != domainparams)
+        return -2;      /* No support for comparing stuff not asked */
+
+    /*
+     * If the two keys belong with different providers or the second key
+     * isn't provided, try to export the second provdata to the first
+     * provider and use that for comparing.
+     */
+    if (keymgmt1 != keymgmt2)
+        provdata2 = exported_provdata2 =
+            evp_keymgmt_export_to_provdata(pkey2, keymgmt1, domainparams);
```

Comment:
IIUC then by calling evp_keymgmt_export_to_provdata directly you avoid all the dirty count checking. But why? Don't we still need to check the dirty counts? What if the underlying legacy key has been amended?

Reply:
The purpose here is to ensure that both keys have cached provdata with the same provider (i.e the same keymgmt), regardless of dirty count. 

However, you're right that I didn't check the dirty count of the first key...

Reply:
Hmmmm, now that you get me looking again, I've been over thinking this, haven't I?  A call to `evp_keymgmt_export_to_provider` would suffice just as well...  BUT, I do need to make that call on `pkey1` as well, just to make sure it's up to date too...

Reply:
Ah, there was a reason after all...  `pkey1` and `pkey2` are constant.
However, we know for a fact that they can only be allocated dynamically, so no matter how much constness we pretend, they aren't really constant, so that might justify bending the rules here.  I'll go with that.
(the flip side of C not having mutables...)

Reply:
This has been completely remade